### DISCUSSION TITLE
Partial Acceptance of Suggestions

### DISCUSSION THREAD

**Discussion by bradykelly** [community/community#12426] [Copilot Conversations]
Often CoPilot will offer a suggestion that is quite a long line of code, but e.g. you only want to accept the first 2/3's of the line, instead of having to accept the whole line then go back and edit it. Some sort of keyboard shortcut would be nice where you can e.g. accept one 'word' of the suggested line at a time and stop when you have the code you want, or still just accept the whole line.

The same thing applies for when CoPilot makes a large, multiline suggestion, and you only want the first so many lines: it would be nice to have another keyboard shortcut that lets you accept the suggestion one line at a time, in addition to being able to accept the whole suggestion. Again, I know I can accept the whole suggestion and edit it, but that slows me down more.

**Comment by PetrDlouhy** [comment]
I already almost started new discussion thread with very similar requirement for `neovim`, so I will copy the text here, because it might illustrate the problem:

Is it possible to make in `neovim` mapping to accept only one character/word from the suggestion?
For example I start to write character `c` and the suggested line turns into:
```python
content_type = ContentType.objects.get_for_model(UUIDArticle, for_concrete_model=False)
```
The beginning of the suggestion is correct, but I want to use different function, for example:
```python
content_type = ContentType.objects.get_content_type()
```

I would like to pres <Ctrl-left> to move one letter in the suggestion ahead and <Ctrl-shift-right> to move one word ahead.

This is inspired by how `fish` terminal treats their suggestions.

**Comment by PetrDlouhy** [comment]
After few hours of experimenting with neoVIM's scripting and mapping I was finally able achieve this without any other support from the Copilot's implementation in neoVIM:
```vim
function! SuggestOneCharacter()
    let suggestion = copilot#Accept("")
    let bar = copilot#TextQueuedForInsertion()
    return bar[0]
endfunction

function! SuggestOneWord()
    let suggestion = copilot#Accept("")
    let bar = copilot#TextQueuedForInsertion()
    return split(bar, '[ .]\zs')[0]
endfunction

imap <script><expr> <C-l> SuggestOneWord()
imap <script><expr> <C-left> SuggestOneCharacter()
```

This maps one word suggestion on Ctrl-l and one character suggestions on Ctrl-left (I was not able to use Ctrl-Shift key combinations).
The word can be separated by space or by `.` character.

**Comment by cmp-nct** [comment]
For IntelliJ IDEs I wrote a plugin that solves the problem: https://github.com/cmp-nct/Stewardess

From working with it for a few days I can see the potential use of Copilot is significantly hindered by the Plugin implementation.
The Plugins give no control, no configuration and they do not provide the backend with the actual code it needs to make proper completion suggestions.
It's clear that the plugins have been hell a lot of work to complete, they include a ton of code. 
It's elusive to me how it could happen that they forgot to "look at next line" when inserting code or "provide SQL CREATE syntax" when working with SQL completions etc.

**Comment by BoD** [comment]
The fact that someone actually took the pain to write a companion plugin just to do that is a testimony on how this feature is so needed!

**Comment by BoD** [comment]
Just want to mention ticket #7331 which seems to have more traction, and is the same idea (even though the title specifically mentions VSCode, whereas this feature would benefit any IDE).

**Comment by Vrajs16** [comment]
According to this: https://github.com/microsoft/vscode/pull/166956#issue-1459807064 , it is already supported in visual studio code extension. Just press ⌘ + →

**Reply by BoD** [reply]
Thanks!
Wow this is very cool for VSCode users, but also now I'm very jealous of them (an an IntelliJ user) :)

**Reply by frederickjjoubert** [reply]
> Thanks!
> Wow this is very cool for VSCode users, but also now I'm very jealous of them (an an IntelliJ user) :)

Same here, I'm sorely missing this feature on CLion and Rider

**Reply by cmp-nct** [reply]
@frederickjjoubert  you can use my above mentioned extension on Jetbrains. Works on all of them, probably even in Appcode.

**Comment by vince91** [comment]
Any update on this? Not sure if it's worth paying for Copilot if I end up rejecting a lot of the suggestions because only the first words are interesting.

**Reply by cmp-nct** [reply]
On VS Code you can use Ctrl+arrow right and left, it works like a charm.
On Jetbrains IDEs you can use my extension (Stewardess) to get the same feature
Not sure about Visual Studio.
You'll end up rejecting 95% of it, that's how copilot works currently as it's based on a very old GPT model but it's still very helpful.

**Comment by bksubramanyarao** [comment]
It's been more than one year, and still for all JetBrains products, I am missing this very important and critical feature, Partial World Accept. When VS Code already has it, why not JetBrains? I don't want to install any other third-party plugin. When I'm paying for software, it looks like the team at Copilot are sleeping.

**Comment by BoD** [comment]
Look what I got after updating today 🎉

<img width="342" alt="Screenshot 2023-11-22 at 16 57 34" src="https://github.com/community/community/assets/372852/f595d204-7a27-4261-9fe2-46ad146137b2">

This ticket can be closed.

**Comment by a-h** [comment]
In `copilot.vim`, there's some commands for this. My setup looks like this:

```lua
-- Copilot setup.
vim.g.copilot_no_tab_map = true
vim.g.copilot_assume_mapped = true
-- Accept the current completion with Ctrl-].
vim.keymap.set('i', '<C-]>', 'copilot#Accept("\\<CR>")', {
  expr = true,
  replace_keycodes = false
})
-- Accept the current completion word with Ctrl-\.
vim.keymap.set('i', '<C-\\>', '<Plug>(copilot-accept-word)')
-- Accept the current completion line with Ctrl-|.
vim.keymap.set('i', '<C-|>', '<Plug>(copilot-accept-line)')
-- Request a suggestion with Ctrl-.
vim.keymap.set('i', '<C-.>', '<Plug>(copilot-suggest)')
```

You can see a list of commands with `:help copilot-maps`
